actors/migration/nv3/miner.go

```diff
@@ -1,27 +1,464 @@
+	}
+
+	var existing power.Claim
+	powerDelta := activePower
+	found, err := claims.Get(abi.AddrKey(a), &existing)
+	if err != nil {
+		return err
+	}
+	if found {
+		powerDelta = powerDelta.Sub(miner.NewPowerPair(existing.RawBytePower, existing.QualityAdjPower))
+	}
+	if err := powSt.AddToClaim(store, a, powerDelta.Raw, powerDelta.QA); err != nil {
+		return err
+	}
+
+	powSt.Claims, err = claims.Root()
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+type expirationQueueStats struct {
+	// total of all active power in the expiration queue
+	totalActivePower miner.PowerPair
+	// total of all faulty power in the expiration queue
+	totalFaultyPower miner.PowerPair
+}
+
+// Updates the expiration queue by correcting any duplicate entries and their fallout.
+// If no changes need to be made cid.Undef will be returned.
+// Returns the new root of the expiration queue
+func (m *minerMigrator) correctExpirationQueue(exq miner.ExpirationQueue, sectors miner.Sectors,
+	allTerminated bitfield.BitField, allFaults bitfield.BitField, sectorSize abi.SectorSize,
+) (cid.Cid, expirationQueueStats, error) {
+	// processed expired sectors includes all terminated and all sectors seen in earlier expiration sets
+	processedExpiredSectors := allTerminated
+	expirationSetPowerSuspect := false
+
+	var exs miner.ExpirationSet
+
+	// Check for faults that need to be erased.
+	// Erased faults will be removed from bitfields and the power will be recomputed
+	// in the subsequent loop.
+	err := exq.ForEach(&exs, func(epoch int64) error { //nolint:nomaprange
+		// Detect sectors that are present in this expiration set as "early", but that
+		// have already terminated or duplicate a prior entry in the queue, and thus will
+		// be terminated before this entry is processed. The sector was rescheduled here
+		// upon fault, but the entry is stale and should not exist.
+		modified := false
+		earlyDuplicates, err := bitfield.IntersectBitField(exs.EarlySectors, processedExpiredSectors)
+		if err != nil {
+			return err
+		} else if empty, err := earlyDuplicates.IsEmpty(); err != nil {
+			return err
+		} else if !empty {
+			modified = true
+			exs.EarlySectors, err = bitfield.SubtractBitField(exs.EarlySectors, earlyDuplicates)
+			if err != nil {
+				return err
+			}
+		}
+
+		// Detect sectors that are terminating on time, but have either already terminated or duplicate
+		// an entry in the queue. The sector might be faulty, but were expiring here anyway so not
+		// rescheduled as "early".
+		onTimeDuplicates, err := bitfield.IntersectBitField(exs.OnTimeSectors, processedExpiredSectors)
+		if err != nil {
+			return err
+		} else if empty, err := onTimeDuplicates.IsEmpty(); err != nil {
+			return err
+		} else if !empty {
+			modified = true
+			exs.OnTimeSectors, err = bitfield.SubtractBitField(exs.OnTimeSectors, onTimeDuplicates)
+			if err != nil {
+				return err
+			}
+		}
+
+		if modified {
+			expirationSetPowerSuspect = true
+			exs2, err := copyES(exs)
```

Comment:
Now that we're not longer putting these in a map for `Set()`ing later, this might not be necessary.